gwcs/wcs.py

```diff
@@ -634,3 +641,235 @@ def fix_inputs(self, fixed):
         new_pipeline.append((step0[0], new_transform))
         new_pipeline.extend(self.pipeline[1:])
         return self.__class__(new_pipeline)
+
+    def to_fits_sip(self, bounding_box, max_error=None, max_rms=None, degree=None,
+                    max_inv_error=None, max_inv_rms=None, verbose=False):
+        """
+        Make a SIP-based approximation to the WCS to be used in FITS WCS
+
+        Parameters
+        ----------
+        bounding_box: a pair of tuples, each consisting of two integers
+            Represents the range of pixel values in both dimensions
+            ((ymin, ymax), (xmin, xmax))
+        max_error: float
+            Maximum allowed error over the domain of the pixel array.
+        max_rms_error: float
+            Maximum allowed rms error over the domain of the pixel array.
+        degree: integer
+            Degree of the SIP polynomial. If supplied, max_error and max_rms_error
+            must be None.
+        max_inv_error: float
+            Maximum allowed inverse error over the domain of the pixel array.
+        max_inv_rms: float
+            Maximum allowed inverse rms over the domain of the pixel array
+        verbose: boolean
+            print progress of fits
+
+        Returns
+        -------
+        fits_dict, max_error, rms_error
+
+        fits_dict is a dictionary where the keys are FITS keywords and the values
+        the corresonding FITS keyword values, to make it easy to merge with
+        an existing header.
+
+        This assumes a tangent projection.
+        """
+        if (self.forward_transform.n_inputs !=2
+            or self.forward_transform.n_outputs != 2):
+           raise ValueError("The to_fits_sip method only works with 2 dimensional transforms")
+
+        # Handle the options related to degree and thresholds.
+        if degree is not None and (max_error is not None or max_rms is not None):
+            raise ValueError(
+                "Degree cannot be specified if either max_error or max_rms have values")
+        if degree is None and max_error is None and max_rms is None:
+            raise ValueError(
+                "At least one of max_error, max_rms, or degree must be specified")
+        fits_dict = {}
+        transform = self.forward_transform
+        # Determine reference points.
+        (ymin, ymax), (xmin, xmax) = bounding_box
+        crpix1 = int((xmax - xmin) / 2)
+        crpix2 = int((ymax - ymin) / 2)
+        crval1, crval2 = transform(crpix1, crpix2)
+        hdr = fits.Header()
+        hdr['naxis'] = 2
+        hdr['naxis1'] = xmax
+        hdr['naxis2'] = ymax
+        hdr['ctype1'] = 'RA---TAN-SIP'
+        hdr['ctype2'] = 'DEC--TAN-SIP'
+        hdr['CRPIX1'] = crpix1
+        hdr['CRPIX2'] = crpix2
```

Comment:
I have two **new** observations for the `CRPIX` issue:

1. This code will probably create a WCS that is shifted by 1 pixel. You _are_ computing SIP correctly (at least as I can see so far), but the FITS WCS is wrong. You are computing and using `CRPIX` in your computations as a `0-based` value but the FITS header expects `1-based` value. Therefore, I think, the simplest fix would be:

```suggestion
        hdr['CRPIX1'] = crpix1 + 1
        hdr['CRPIX2'] = crpix2 + 1
```

2. Because you did not caught it with your current test, I would suggest adding a test specifically designed for this: when at `CRPIX` wcs transformation should return `CRVAL`. Essentially you want to check (`allclose`):
```python
w.wcs_pix2world(*w.wcs.crpix, 1) == w.wcs.crval  # allclose instead of equality
```

Reply:
I had tried that previously but had gotten wrong results, which I found surprising. I'll check again.

Reply:
Pay attention to the `origin` parameter (3rd parameter) of the `wcs_pix2world` - it should be 1. If it is 0 then you should supply `crpix-1` as input.

Reply:
What I meant in my last comment, is that if you _prefer_ using 0-based values, then the test that _must pass_ is:
```python
assert np.allclose(w.wcs_pix2world(*(w.wcs.crpix - 1), 0), w.wcs.crval, atol=0, rtol=1e-9)
```
**ASSUMING** that you do set `CRPIX1` and `CRPIX2` as 1-based value, i.e.,
```
        hdr['CRPIX1'] = crpix1 + 1
        hdr['CRPIX2'] = crpix2 + 1
```

Reply:
I've figured out how to get it right.  By the way, the test you suggested passed in the previous version (I did it by hand several times) so it would not have found this problem.  I must admit I had been puzzled by the origin issue and wondered if I misunderstood how it worked.